### PR TITLE
feat(relay): Add public_key to RelayUsage

### PR DESCRIPTION
**Pull request by RaduW**
**Note:** sentry_relayusage table is virtually empty (probably under 10 records currently), and it is very
rarely accessed (probably a few times a minute) so there shouldn't be any major performance implications when
adding an index on the column `public_key`.

```sql
BEGIN;
--
-- Add field public_key to relayusage
--
ALTER TABLE "sentry_relayusage" ADD COLUMN "public_key" varchar(200) DEFAULT '' NOT NULL;
ALTER TABLE "sentry_relayusage" ALTER COLUMN "public_key" DROP DEFAULT;
CREATE INDEX "sentry_relayusage_public_key_d2a87c66" ON "sentry_relayusage" ("public_key");
CREATE INDEX "sentry_relayusage_public_key_d2a87c66_like" ON "sentry_relayusage" ("public_key" varchar_pattern_ops);
COMMIT;
```

### CODE REVIEW

File: src/sentry/migrations/0104_collect_relay_public_key_usage.py
Comment (untitaker) [inline src/sentry/migrations/0104_collect_relay_public_key_usage.py:24]: when redash is up again let's verify before merge that this table is basically empty (I think it is). otherwise this won't fly

**Review by untitaker — APPROVED** [review]

File: src/sentry/migrations/0104_collect_relay_public_key_usage.py
Comment (jan-auer) [reply src/sentry/migrations/0104_collect_relay_public_key_usage.py:24]: I would assume that we want both `CREATE INDEX CONCURRENTLY` and set this to false, as the comment indicates.

File: src/sentry/migrations/0104_collect_relay_public_key_usage.py
Comment (jan-auer) [inline src/sentry/migrations/0104_collect_relay_public_key_usage.py:18]: This was coordinated with operations. The table has just been inserted and is practically empty. Onpremise and ST will ship this migration alongside the previous one. Under these circumstances, this migration does not need to be marked `is_dangerous` 👍

File: src/sentry/models/relay.py
Comment (jan-auer) [inline src/sentry/models/relay.py:20]: This does not create a default in SQL, meaning that the deploy will break. Either, we need to enforce a DB default, or we need to make this nullable.

**Review by jan-auer — CHANGES_REQUESTED** [review]

File: src/sentry/migrations/0104_collect_relay_public_key_usage.py
Comment (untitaker) [reply src/sentry/migrations/0104_collect_relay_public_key_usage.py:24]: let's wait for owners-migrations but I do think this is fine. `CREATE INDEX CONCURRENTLY` involves hand-written sql afaik and I don't think we need that stuff for an empty table...

File: src/sentry/models/relay.py
Comment (RaduW) [reply src/sentry/models/relay.py:20]: While this is true, in practice I think we can live with it.

A registration error will happen for the cases when a new Relay tries to register exactly between the application of the migrations and the recycling of the web worker. 

For existing Relays that register the update should be fine even with the old model accessing the new schema.
So we only have a  small time when new relays are temporarily unable to register themselves.

File: src/sentry/migrations/0104_collect_relay_public_key_usage.py
Comment (markstory) [reply src/sentry/migrations/0104_collect_relay_public_key_usage.py:24]: I took a look in prod and there are only a handful of rows in this table. You don't need to create the index concurrently with a low row count and low read/write traffic.

**Review by markstory — APPROVED** [review]
Looks reasonable to me.

**Comment by untitaker** [comment]
I think we need to address the nullability/default situation then this should be good for merging. I would not really bet on a race during deployment.

File: src/sentry/models/relay.py
Comment (wedamija) [inline src/sentry/models/relay.py:20]: We typically recommend that you just use a `models.TextField` here. This just helps you avoid a painful migration if you decide to change the column length at some point in the future. If you want to enforce a length in Django, @mattrobenolt wrote an alternative `CharField` that is backed by a `text` column in postgres: https://github.com/getsentry/sentry/blob/3b5490df3b37ce20a3d01ae3b7c1114ee1e00c3f/src/sentry/db/models/fields/text.py#L28

Comment (wedamija) [reply src/sentry/models/relay.py:20]: I'll leave this to your judgement, but do you even need a default of an empty string here? Could we just make this column nullable instead of a blank default?

Comment (wedamija) [reply src/sentry/models/relay.py:20]: If we want to do this safely and keep the string default, my recommendation would be:
1. Change this column to be nullable with a default of "". Deploy
2. Add a new migration to make the column not null. I believe that Django will re-apply the default to the table to catch any other null values, but we can check the generated DDL to be sure in that pr.

Comment (jan-auer) [reply src/sentry/models/relay.py:20]: @wedamija We'd probably want to keep this consistent with the other model in that file that uses a CharField for `public_key`, too. WDYT?

**Comment by RaduW** [comment]
@wedamija, thanks for your suggestion. 

The public_key will never be empty/null for new Relays or after the existing relay entries are updated (when the relays do their re-register requests) so I'd rather go with a temporary empty default.

File: src/sentry/api/endpoints/relay_register.py
Comment (jan-auer) [inline src/sentry/api/endpoints/relay_register.py:198]: Shouldn't this rather set the public_key during creation? Otherwise, you're creating usage entries without a key, but you're updating the key every time where it can never change.

```suggestion
            relay_usage = RelayUsage.objects.get(relay_id=relay_id, version=version)
        except RelayUsage.DoesNotExist:
            RelayUsage.objects.create(relay_id=relay_id, version=version, public_key=public_key)
        else:
            relay_usage.last_seen = timezone.now()
```

**Review by jan-auer — COMMENTED** [review]
One more thing: We should probably debounce writes to `last_seen`. There was code in the past that used a cache to only update once per minute. I'll dig that out.

**Edit**: Seems like I did not remember correctly. Feel free to leave as-is.

File: src/sentry/api/endpoints/relay_register.py
Comment (RaduW) [reply src/sentry/api/endpoints/relay_register.py]: right, the public_key argument from the getter should be in the create, sorry

**Review by jan-auer — APPROVED** [review]
Thanks, @RaduW

File: src/sentry/models/relay.py
Comment (wedamija) [reply src/sentry/models/relay.py]: So I did a bit of reading of this, and it might actually no longer be a problem to change the lengths of varchar cols: https://stackoverflow.com/questions/48693373/does-alter-column-type-varcharn-rewrite-the-table-in-postgres-9-6.

Really up to y'all about how likely you are to need to change the length of this column. IMO, there's not a lot of reason to use a varchar in most cases, but I'm not going to block on it or anything.

Comment (jan-auer) [reply src/sentry/models/relay.py]: Ok, thanks. My understanding is that for a while VARCHAR has effectively been implemented as TEXT in pg, but I'd have to go back and check. The only reason for me to keep it as CharField now is consistency with other models that have the same `public_key`.

Comment (wedamija) [reply src/sentry/models/relay.py]: Yeah, you're right that it's a `TEXT` column under the hood. It used to be the case that changing the length would cause a table rewrite, which is not great in production. This may have changed though.

**Pull request merged**
